lib/internal/vm/noderealm.js

```diff
@@ -0,0 +1,132 @@
+'use strict';
+
+// NodeRealm was originally a separate module developed by
+// Anna Henningsen and published separately on npm as the
+// synchronous-worker module under the MIT license. It has been
+// incorporated into Node.js with Anna's permission.
+// See the LICENSE file for LICENSE and copyright attribution.
+
+const {
+  Promise,
+} = primordials;
+
+const {
+  NodeRealm: NodeRealmImpl,
+} = internalBinding('contextify');
+
+const EventEmitter = require('events');
+const { setTimeout } = require('timers');
+const { dirname, join } = require('path');
+
+let debug = require('internal/util/debuglog').debuglog('noderealm', (fn) => {
+  debug = fn;
+});
+
+class NodeRealm extends EventEmitter {
+  #handle = undefined;
+  #process = undefined;
+  #global = undefined;
+  #module = undefined;
+  #stoppedPromise = undefined;
+
+  /**
+   */
+  constructor() {
+    super();
+    this.#handle = new NodeRealmImpl();
+    this.#handle.onexit = (code) => {
+      this.stop();
+      this.emit('exit', code);
+    };
+    try {
+      this.#handle.start();
+      this.#handle.load((process, nativeRequire, globalThis) => {
+        this.#process = process;
+        this.#module = nativeRequire('module');
+        this.#global = globalThis;
+        process.on('uncaughtException', (err) => {
+          if (process.listenerCount('uncaughtException') === 1) {
+            // If we are stopping, silence all errors
+            if (!this.#stoppedPromise) {
+              this.emit('error', err);
+            }
+            process.exit(1);
+          }
+        });
+      });
+    } catch (err) {
+      this.#handle.stop();
+      throw err;
+    }
+  }
+
+  /**
+   * @returns {Promise<void>}
+   */
+  async stop() {
+    // TODO(@mcollina): add support for AbortController, we want to abort this,
+    // or add a timeout.
+    return this.#stoppedPromise ??= new Promise((resolve) => {
+      const tryClosing = () => {
+        const closed = this.#handle.tryCloseAllHandles();
+        debug('closed %d handles', closed);
+        if (closed > 0) {
+          // This is an active wait for the handles to close.
+          // We might want to change this in the future to use a callback,
+          // but at this point it seems like a premature optimization.
+          // TODO(@mcollina): refactor to use a close callback
+          setTimeout(tryClosing, 100);
```

Comment:
The choice of 100ms for the timeouts here and on line 89 seem kind of arbitrary.